nomad/structs/structs.go

```diff
@@ -10273,6 +10279,40 @@ func (a *Allocation) Reconnected() (bool, bool) {
 	return true, a.Expired(lastReconnect)
 }
 
+func (a *Allocation) ToIdentityClaims() *IdentityClaims {
+	now := jwt.NewNumericDate(time.Now().UTC())
+	return &IdentityClaims{
+		Namespace:    a.Namespace,
+		AllocationID: a.ID,
+		RegisteredClaims: jwt.RegisteredClaims{
+			// TODO: in Nomad 1.5.0 we'll have a refresh loop to
+			// prevent allocation identities from expiring before the
+			// allocation is terminal. Once that's implemented, add an
+			// ExpiresAt here ExpiresAt: &jwt.NumericDate{},
+			NotBefore: now,
+			IssuedAt:  now,
+		},
+	}
+}
+
+func (a *Allocation) ToTaskIdentityClaims(taskName string) *IdentityClaims {
+	claims := a.ToIdentityClaims()
+	if claims != nil {
+		claims.TaskName = taskName
+	}
+	return claims
+}
+
+// IdentityClaims are the input to a JWT identifying a workload. It
+// should never be serialized to msgpack unsigned.
+type IdentityClaims struct {
+	Namespace    string `json:"nomad_namespace"`
+	AllocationID string `json:"nomad_allocation_id"`
+	TaskName     string `json:"nomad_task"`
```

Comment:
Hm, I wonder how much we need to get these right on the first try since they're going to live as long as their allocation and allocation's might live a long time. (An argument for expiration I guess!)

If we want to be safe I think `job` will be necessary for user services to find them useful to do auth like:

> `Task api of Job user-api` is allowed to write to `Task db of Job user-db`

Where `write` is a capability encoded in the user's code doing the jwt-based auth.

I suspect region might also be useful:

> `Task api of Job user-api in Region Z` is allowed to write to `Task db of Job user-db in Region Z`

Unsure about that one though.

At one point I was thinking ClusterID could even be included to ensure claims are universally unique, but then I realized that's the job of the signing key. I don't think we want people trying to validate the signing key as well as the ClusterID (which we don't generally expose anyway) to ensure the JWT is minted by their trusted Nomad cluster. I don't think ClusterID would add anything useful.

Reply:
Added JobID in 23af20e29b2398437d86c45487d7bccf7f639cb3